### PR TITLE
CIP-0086? | NFT Metadata Update Oracles

### PR DESCRIPTION
**Pull request by nicolasayotte**
([current proposal draft in branch](https://github.com/nicolasayotte/CIPs/blob/nft-metadata-oracle/CIP-0086/README.md))

### CODE REVIEW

**Comment by mathieu2em** [comment]
Definitely a useful feature that would come handy for a lot of Cardano NFT projects. As someone working with a lot of them I can already view many different places where this would be useful. Thank you so much for the work done here ! +1 love to the regex integration.

**Comment by MicroProofs** [comment]
I think you should incorporate CIP-67 (the label standard). This would help with offchain being able to easily identify tokens following this standard. CIP-67 can be updated to include a label to reflect the usage of this standard.

**Comment by GeorgeFlerovsky** [comment]
> I think you should incorporate CIP-67 (the label standard). This would help with offchain being able to easily identify tokens following this standard. CIP-67 can be updated to include a label to reflect the usage of this standard.

Thanks for your suggestion. We'll check it out!

**Comment by nicolasayotte** [comment]
> I think you should incorporate CIP-67 (the label standard). This would help with offchain being able to easily identify tokens following this standard. CIP-67 can be updated to include a label to reflect the usage of this standard.

As such, our standard is agnostic to the exact asset names representation and, from what I can surmise, our CIP technically supports CIP-67 as-is.

CIP-67 is an interesting concept: maybe its registry should include a label for CIP-25 standard NFTs?

**Comment by the-Minister-0001** [comment]
Hey Nicolas, thanks for sharing this!
I think the functionality you described would enrich a lot of projects, let alone the one you're working on.

Nice to see you formulating out what you're doing for your fairly large scale application! Hopefully it'll benefit other projects, too!

**Comment by huths0lo** [comment]
This is a rather novel approach to what will become a significant problem in the future.  I've personally had to overcome the challenge of updating metadata on the Cardano Budz GYOA platform.  Since no other option exists, I went the route of reminting the assets every time an update is needed.  But I did this knowing that there was major downsides, with no alternative solution:

1) Cost for reminting (Shouldered by the project owner, or asset owner)
2) Blockchain bloat
3) Additional complexity for maintaining NFT status by ensuring the extra token created is burned
4) Not being able to use a locking policy on the assets

I think this is a very novel approach to solving these inherent problems of reminting tokens.  There is no overly complex approach to proving you are the original owner, as you are repurposing the policy skey as a Cardano Wallet.  It addresses that some NFT projects may want to prove they will never remint their assets by providing an on chain mechanism.  It addresses the burden of having to index these updates for general use.  It addresses using regex to allow for mass updates.  And it simplifies the updating of many issues in a singular transaction, instead of many.  Although you could update many tokens already, this makes it much simpler since the updates wouldnt require additional asset management.

**Comment by SmaugPool** [comment]
Hi, 

What about CIP25 version 2 where the policy ID and asset name are encoded as byte strings?

The asset name may not be decodable as UTF-8 so how would the asset name in updates be specified and how would regex updates work?

**Comment by GeorgeFlerovsky** [comment]
> Hi, 
> 
> What about CIP25 version 2 where the policy ID and asset name are encoded as byte strings?
> 
> The asset name may not be decodable as UTF-8 so how would the asset name in updates be specified and how would regex & tabular updates work?

Good point about non UTF-8 token names. I guess the easy answer to that would be to disqualify such tokens (with non-text token names) from being updatable by our CIP's regex and tabular mechanisms.

They might still be eligible for our simple update mechanism, so they would still benefit from our CIP because they wouldn't need to remint to update token metadata.

Do you have an alternative suggestion to handle this?

**Comment by SmaugPool** [comment]
> They might still be eligible for our simple update mechanism

How? It seems that the asset name is also a string in the simple update mechanism.

As an aside one benefit of CIP25v2 beyond allowing non-text asset names is that byte strings are half as long as hex strings so this saves some space in the blockchain for the policy ID at least. Maybe this could have been considered for this CIP too.

File: CIP-????/README.md [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+
+## Updating an oracle assignment
+
+The metadata oracle assignment for a policy ID can be updated via a transaction signed by the oracle main address key. The transaction must only send ADA from the oracle main address to itself and must not mint any tokens, but it may contain any number of inputs and outputs. Otherwise, the transaction is ignored for the purposes of metadata oracle assignment.
+
+The schema for updating an oracle assignment is the same as for the initial assignment in the minting transaction:
+
+```json
+{
+    "????": {
```
Comment (SebastienGllmt) [inline CIP-????/README.md:94]: Can you add cddl definitions for all of these? Since those are what are actually stored on the ledger

Comment (nicolasayotte) [reply CIP-????/README.md:94]: Good idea! Thanks

Comment (GeorgeFlerovsky) [reply CIP-????/README.md:94]: Yes, certainly. We'll add the CDDL definitions for everything, this week.

**Comment by GeorgeFlerovsky** [comment]
> > They might still be eligible for our simple update mechanism
> 
> How? It seems that the asset name is also a string in the simple update mechanism.
> 
> As an aside one benefit of CIP25v2 beyond allowing non-text asset names is that byte strings are half as long as hex strings so this saves some space in the blockchain for the policy ID at least. Maybe this could have been considered for this CIP too.

Ah, right. We would indeed have to use an alternative to text-based JSON field names if we want to support non-text asset names. I'll check out CIP-25-v2 to see if we can adapt their approach to our CIP.

**Comment by nicolasayotte** [comment]
> > They might still be eligible for our simple update mechanism
> 
> How? It seems that the asset name is also a string in the simple update mechanism.
> 
> As an aside one benefit of CIP25v2 beyond allowing non-text asset names is that byte strings are half as long as hex strings so this saves some space in the blockchain for the policy ID at least. Maybe this could have been considered for this CIP too.

I am sure we can have the policy id and asset name level tag support CIP-25-v2 with a version tag at least for simple updates. For the bulk updates, I will have to think about how / if those can be implemented.

File: CIP-????/README.md (hunk 1 above)
Comment (nicolasayotte) [reply CIP-????/README.md:94]: @GeorgeFlerovsky I am working on it, btw. :D

**Comment by rphair** [comment]
thanks @GeorgeFlerovsky for https://github.com/cardano-foundation/CIPs/issues/439#issuecomment-1386027179 and you are welcome to add the H1 and make the other sections H2. 🤓 The exact (longer) titles for the _Motivation_  and _Rationale_ sections are listed in https://github.com/cardano-foundation/CIPs/tree/master/CIP-0001#structure (see also background at https://github.com/cardano-foundation/CIPs/issues/439#issuecomment-1386599115)

File: CIP-????/README.md [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+
+For each oracle main address currently assigned to a policy ID, the indexer monitors the blockchain for non-minting transactions that only send ADA from the oracle main address to itself and contain oracle assignment updates in their metadata. The indexer updates its database to reflect these explicit oracle assignments and removes any implicit assignments that were replaced by explicit assignments.
+
+For each oracle update address currently assigned to a policy ID, the indexer monitors the blockchain for non-minting transactions that only send ADA from the oracle update address to itself and contain CIP-???? token metadata updates. The indexer applies these metadata updates in the order defined in [Order of application for updates](https://www.notion.so/Order-of-application-for-updates-07b8a74001c9484698f400f9c5e1c7c8). CIP-???? metadata updates are applied to the asset classes and metadata fields that they target, while keeping all other fields the same.
```
Comment (ross-spencer) [inline CIP-????/README.md:303]: Lines 303, 378, and 379, link to notion.so. Should they be relative links within this text?

File: CIP-????/README.md [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+Title: NFT Metadata Update Oracles
+Status: Proposed
+Category: Metadata
+Authors:
+    - Nicolas Ayotte <nick@equine.gg>
+    - George Flerovsky <george@mlabs.city>
+    - Samuel Williams <samuel@mlabs.city>
+Implementors:
+    - Equine <[https://www.equine.gg/](https://www.equine.gg/)>
+    - MLabs <[https://mlabs.city/](https://mlabs.city/)>
```
Comment (rphair) [inline CIP-????/README.md:12]:
```suggestion
    - Equine <https://www.equine.gg/>
    - MLabs <https://mlabs.city/>
```
(markdown doesn't cause an error within the preamble fields, but it's not recognised either)

File: CIP-????/README.md [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+
+## Tabular metadata updates
+
+Tabular metadata updates use a condensed rectangular format to specify new values for a fixed set of fields for a large number of assets. Specifically, we use the comma-separated values (CSV) format:
+
+```json
+{
+  "????": {
+	  "tabular_metadata_update": {
+	    "<policyId>": "<CsvValue>"
```
Comment (rphair) [inline CIP-????/README.md:181]:
```suggestion
	  "tabular_metadata_update": {
	    "<policyId>": "<CsvValue>"
	    }
```
The missing closure here makes it think that the text after the JSON block is still JSON and is also highlighted as errors.  It renders properly on GitHub but might not in other environments.

**Review by rphair — COMMENTED** [review]
@GeorgeFlerovsky in your next revision, including
- (demote top level headers to H2) https://github.com/cardano-foundation/CIPs/pull/430#issuecomment-1386607038
- (section links point to notion.so instead of document) https://github.com/cardano-foundation/CIPs/pull/430#discussion_r1084572077

can you also...

File: CIP-????/README.md (hunk 1 above)
Comment (GeorgeFlerovsky) [reply CIP-????/README.md:95]: @sebastiengllmt @smaugpool
We've added CDDL files to our proposal.

Similar to CIP-25, we have:

- `cddl/version_1.cddl` for text-based policy IDs and asset names.
- `cddl/version_2.cddl` for bytes-based policy IDs and asset names.

File: CIP-????/README.md (hunk 2 above)
Comment (GeorgeFlerovsky) [reply CIP-????/README.md]: Fixed

File: CIP-????/README.md (hunk 3 above)
Comment (GeorgeFlerovsky) [reply CIP-????/README.md]: Fixed

File: CIP-????/README.md (hunk 4 above)
Comment (GeorgeFlerovsky) [reply CIP-????/README.md:182]: Fixed

**Comment by rphair** [comment]
@alessandrokonrad we were just saying in the CIP meeting now that maybe you could give a review of this?

**Comment by alessandrokonrad** [comment]
Does it need to be fully backwards compatible to CIP-0025? As it was pointed out in several other CIP discussions CIP-0025 is flawed and allows for spoofing attacks. Forcing asset names to have a label (CIP-0067) that attack vector significantly. And I think this CIP should consider this and only recognize CIP-0025 assets that follow  a specific label.

File: CIP-????/README.md [hunk 5]
```diff
@@ -0,0 +1,396 @@
+---
+CIP: ????
```
Comment (KtorZ) [inline CIP-????/README.md:2]:
```suggestion
CIP: 86
```

File: CIP-????/README.md [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+    - Samuel Williams <samuel@mlabs.city>
+Implementors:
+    - Equine <https://www.equine.gg/>
+    - MLabs <https://mlabs.city/>
+Discussions:
+    - https://github.com/cardano-foundation/CIPs/pull/430
+Created: 2022-11-01
+License: CC-BY-4.0
+---
+# CIP-????: NFT Metadata Update Oracles
```
Comment (KtorZ) [inline CIP-????/README.md:18]:
```suggestion
```

File: CIP-????/README.md [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+
+## Abstract
+
+This proposal extends the CIP-25 standard for defining and updating token metadata via transaction metadata, by providing a new mechanism to update token metadata without having to mint or burn tokens, while maintaining full backward compatibility with CIP-25. The new mechanism is capable of expressing metadata updates more efficiently than CIP-25 updates.
+
+## Motivation
```
Comment (KtorZ) [inline CIP-????/README.md:24]:
```suggestion
## Motivation: why is this CIP necessary?
```

File: CIP-????/README.md [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+
+For tabular metadata updates, the bytestring CSV value may get broken up into an array of bytestring chunks in the CBOR representation. When this happens, the indexer recombines the chunks into the whole CSV table before applying the tabular metadata update.
+
+To be able to handle blockchain rollbacks, the indexer keeps track of past metadata states for its policy IDs, going back 2160 blocks (~12 hours) from the current blockchain tip. Cardano’s `securityParam` Shelley Genesis parameter prevents nodes from rolling back more than 2160 blocks. If the Cardano node informs the indexer of a rollback, the indexer restores the past metadata state that existed immediately before all the metadata updates in the rolled-back blocks were applied.
+
+For compatibility with existing applications that are already relying on CIP-25 metadata indexers, the CIP-???? indexer provides a similar API so that those applications can get and display the current CIP-???? token metadata in the same way that they have been for CIP-25 metadata. The indexer indicates that it is following the CIP-???? standard.
+
+## Rationale
```
Comment (KtorZ) [inline CIP-????/README.md:313]:
```suggestion
## Rationale: how does this CIP achieve its goals?
```

File: CIP-????/README.md (hunk 1 above)
Comment (nicolasayotte) [reply CIP-????/README.md:94]: @SebastienGllmt we had added the definition files

**Comment by nicolasayotte** [comment]
> Does it need to be fully backwards compatible to CIP-0025? As it was pointed out in several other CIP discussions CIP-0025 is flawed and allows for spoofing attacks. Forcing asset names to have a label (CIP-0067) reduce that attack vector significantly. And I think this CIP should consider this and only recognize CIP-0025 assets that follow a specific label.

Well, it doesn't **need** to be fully backwards compatible, but it would more enabling for all the early projects that promised gamification (especially those that are now policy locked and cannot adapt their NFTs) if we are. As you are well aware, innovative NFT standards exist that provide higher functionality for the future of Cardano (like CIP-68). CIP-86 is tailored for projects that either already launched their collection, or do not want (or are unable) to switch to Smart Contracts development to manage all the possible states of their NFTs.

**Review by rphair — COMMENTED** [review]
@nicolasayotte @GeorgeFlerovsky looks ready to go after last round of edits, with almost all feedback incorporated... it sounds to me like https://github.com/cardano-foundation/CIPs/pull/430#issuecomment-1434653822 has an indirect answer in the [Token metadata indexer](https://github.com/nicolasayotte/CIPs/blob/nft-metadata-oracle/CIP-%3F%3F%3F%3F/README.md#token-metadata-indexer) section (?).

Just please first rename folder `CIP-????` to `CIP-0086` and then I'm happy to approve this 🤓

**Comment by nicolasayotte** [comment]
> @nicolasayotte @GeorgeFlerovsky looks ready to go after last round of edits, with almost all feedback incorporated... it sounds to me like [#430 (comment)](https://github.com/cardano-foundation/CIPs/pull/430#issuecomment-1434653822) has an indirect answer in the [Token metadata indexer](https://github.com/nicolasayotte/CIPs/blob/nft-metadata-oracle/CIP-%3F%3F%3F%3F/README.md#token-metadata-indexer) section (?).
> 
> Just please first rename folder `CIP-????` to `CIP-0086` and then I'm happy to approve this 🤓

I rename the folder! 🎉

**Review by rphair — COMMENTED** [review]
yes @nicolasayotte it's been renamed... but with some files (`README.md`, `*.cddl`) now duplicated into a `CIP-????` directory 😖 ... after this is fixed we should also check the links to subordinate files from `README.md` because of the duplication.

**Comment by nicolasayotte** [comment]
> yes @nicolasayotte it's been renamed... but with some files (`README.md`, `*.cddl`) now duplicated into a `CIP-????` directory 😖 ... after this is fixed we should also check the links to subordinate files from `README.md` because of the duplication.

I will try again: "?" in a path on Windows simply doesn't work at all, so I am going through WSL to do this 😝

**Comment by nicolasayotte** [comment]
@rphair fixed correctly this time. 👍

**Review by rphair — APPROVED** [review]

File: CIP-0086/README.md [hunk 9]
```diff
@@ -0,0 +1,793 @@
...
+
+Our proposal extends CIP-25 with a new update mechanism:
+
+-   A metadata oracle can be assigned for a given policy ID.
+-   For a policy with a metadata oracle assigned,
+the metadata oracle can post CIP-86 updates to add, remove, or modify
+any fields of the token’s CIP-25 metadata,
+without needing to mint or burn any tokens in the metadata update transactions.
+-   The current metadata for a token can be deterministically reconstructed
+by starting from the latest CIP-25 update to the token,
```
Comment (euonymos) [inline CIP-0086/README.md:123]: should be: by starting from the **earliest**

**Comment by nftguild** [comment]
@nicolasayotte , the main contributor to CIP 86, gives an overview of what the CIP is all about and what it’s trying to accomplish, in this short video:

**Intro to CIP 86 - NFT Metadata Update Oracles**
https://youtu.be/HN40XybkAKs 

You can also watch the entire NFT Guild roundtable discussion:
https://youtu.be/Du3Csxwt7cU

Transcript: https://drive.google.com/file/d/1L6cz1PlBWK5bKCLCr2YytDIJypynIiC7/view?usp=share_link

**Comment by rphair** [comment]
@KtorZ I think this will be a good inclusion for Review at our next CIP meeting (currently scheduled for 02 May) since I believe the author has addressed all feedback adequately and there appears to be community support for this. 🤩

**Comment by rphair** [comment]
resounding approval at CIP meeting just ended now... @KtorZ it sounded like you were happy with it; would you like to do 1 more green-check review & merge? 🤠

**Review by Ryun1 — APPROVED** [review]
Following todays meeting, it appears all community concerns have been addressed by the authors and there is no significant disagreement stopping this from moving to be merged as proposed.

**Pull request merged**